instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb

```diff
@@ -87,6 +93,14 @@ def annotate_span_with_response!(span, response)
             def tracer
               Net::HTTP::Instrumentation.instance.tracer
             end
+
+            def untraced?
+              return true if instrumentation_config[:untraced_hosts]&.any? do |host|
+                host === @address # rubocop:disable Style/CaseEquality
+              end
+
+              false
+            end
```

Comment:
```suggestion
            def untraced?
              instrumentation_config[:untraced_hosts].include?(@address)
            end
```

We should avoid `===`, since its uses are [far more interesting and varied](https://medium.com/rubyinside/triple-equals-black-magic-d934936a6379) than what we want here.

Given that the default for the option is `[]` and we are validating that it is an Array, I think we can simplify the entire method to be an `includes?` check.

Reply:
The intention here is to support both String and Regexp. Your proposed change doesn't work:

```ruby
[/foo/].include?('foo') #=> false
```

Reply:
This PR didn't say that you were asking it to validate regular expressions as well. I don't know if we _should_ allow that. Triple-equals allows a lot of loopholes.

In the context of "this is an array of strings", it would work fine.

Reply:
We should allow Regexp. An example use case is to include a domain and its subdomains:

```
/\A(.*\.)?google\.com\z/
```

Reply:
Done.

Reply:
I'm on the fence about it myself - I have the vaguest recollection of us deciding not to allow regular expressions before. Maybe @fbogsany or @robertlaurin can remember?

If they don't recall a specific thing, then perhaps it was a fever dream.

Reply:
I don't remember specifically not allowing regular expressions.

If I remember correctly for rack that case was made that we didn't want to add a config option for everyones preferred approach of filtering. [We ended up with two config options](https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/main/instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/tracer_middleware.rb#L95-L100), the first pass pass which simply accepted an array of strings representing endpoints to omit, and the ultra malleable lambda. 

I think this PR can go a couple of directions.

- We could say this should be solved via a sampler and drop this PR entirely, I've seen this approach more widely pushed in the instrumentation sigs whenever this subject has come up. 

That would look something like the (untested) snippet below, the benefit here is that not every http instrumentation gem needs to be updated. So it no longer becomes a concern of the OpenTelemetry maintainers as this approach is already supported.
```rb
class CustomSampler
  attr_reader :description

  RESULT = OpenTelemetry::SDK::Trace::Samplers::Result
  DECISION = OpenTelemetry::SDK::Trace::Samplers::Decision

  def initialize(untraced_hosts:, description:)
    @untraced_hosts = untraced_hosts
    @description = description
  end

  def ==(other)
    @untraced_hosts == other.untraced_hosts && @description == other.description
  end

  # @api private
  #
  # See {Samplers}.
  def should_sample?(trace_id:, parent_context:, links:, name:, kind:, attributes:)
    sample_decision = if kind == :client && untraced?(attributes['net.peer.name'])
      DECISION::DROP
    else
      DECISION::RECORD_AND_SAMPLE
    end

    RESULT.new(decision: sample_decision, tracestate: OpenTelemetry::Trace.current_span(parent_context).context.tracestate)
  end

  def untraced?(host)
    return true if @untraced_hosts&.any? do |untraced_host|
      host.is_a?(Regexp) ? host.match?(host) : untraced_host == host
    end

    false
  end
end
```

- If we decide that the sampler approach is not suitable for whatever reason, we could consider amending this PR to make it consistent with the [rack configuration options ](https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/main/instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/tracer_middleware.rb#L95-L100). Rack and http client gems have the common ground of the http semantic conventions, so it makes sense to me for the configuration options to have some parity.

- We could deviate to something a bit different, like what is proposed in this PR.

@johnnyshields you've been an active adopter and contributor to otel ruby for a while now, I'm interested to know what you think of the sampler approach for this instead of adding a config options.

Reply:
> We could consider amending this PR to make it consistent with the [rack configuration options](https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/main/instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/tracer_middleware.rb#L95-L100)

I actually originally a lambda option in this PR like Rack, but due to the gnarly specifics of `Net::HTTP`, it just doesn't feel as clean--you'd need 3 lambda args for `@address`, request, and request body respectively. (Rack *happens* to be designed for pluggability, which one `env` variable that is passed thru your middleware stack, and also into the lambda.)

I also couldn't think of a clear use case for the lambda that I couldn't address with Regexp. So I thought it would be better to implement in a separate PR if/when someone actually needs it.

> We could say this should be solved via a sampler and drop this PR entirely

Hmm... I think filtering out specific domains vs. "sampling" a subset of your overall requests are two different concerns/use cases, at least in my mind. It seems to be overkill to have to create a sampler class when I really just want to ignore one or two domains on this specific Net::HTTP instrumenter. Moreover, I'd then have to worry about chaining my "ignore domain" sampler with my overall "1-out-of-1000 requests" sampler, or else kludge the two concerns together. It also seems that OTEL Ruby currently [lacks basic programmatic configurability of samplers](https://github.com/open-telemetry/opentelemetry-ruby/issues/321#issuecomment-1189599314).

**TLDR;** I think we should merge this (minimal) PR as-is and defer decision for both Rack-style lambda arg and samplers.